share/normalize/links.py

```diff
@@ -583,13 +677,111 @@ class DOILink(AbstractLink):
 
-    def execute(self, obj):
-        if not isinstance(obj, str):
-            raise TypeError('{} is not of type str'.format(obj))
+    @classmethod
+    def hint(cls, obj):
+        return int('10.' in obj) * .5 + int('doi' in obj) * .5
+
+    def _process_scheme(self, _):
+        return self.DOI_SCHEME
+
+    def _process_authority(self, _):
+        return self.DOI_DOMAIN
+
+    def _parse(self, obj):
         match = re.search(self.DOI_RE, obj.upper())
         if not match:
-            raise ValueError('{} is not a valid DOI'.format(obj))
-        return furl.furl('{}{}'.format(self.DOI_URL, *match.groups())).url
+            raise ValueError('\'{}\' is not a valid DOI.'.format(obj))
+        return {
+            'scheme': None,
+            'authority': None,
+            'path': '/' + '/'.join(urllib.parse.quote(x, safe=self.SAFE_SEGMENT_CHARS) for x in match.group().split('/'))
+        }
+
+
+class URLLink(AbstractIRILink):
+    PORTS = {80, 443, 20, 989}
+    SCHEMES = {'http', 'https', 'ftp', 'ftps'}
+    URL_RE = re.compile(r'[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)')
+
+    @classmethod
+    def hint(cls, obj):
+        if cls.URL_RE.search(obj) is not None:
+            return 0.25
+        return 0
+
+    def _process_scheme(self, scheme):
+        scheme = scheme.lower()
+        if scheme not in self.SCHEMES:
+            raise ValueError('\'{}\' is not a valid scheme for URLs.'.format(scheme))
+        return scheme.rstrip('s')  # Standardize on non-secure
+
+    def _process_query(self, query):
+        return query  # TODO Order me
+
+    def _process_authority(self, authority):
+        authority = super()._process_authority(authority)
+        if ':' in authority:
+            authority, port = authority.split(':')
+        else:
+            port = None
+        if port and int(port) not in self.PORTS:
+            raise ValueError('\'{}\' is not a valid port for URLs.'.format(port))
+        return authority
+
+
+class EmailLink(AbstractIRILink):
+
+    EMAIL_RE = re.compile(r'(?P<scheme>mailto:)?(?P<mailbox>[a-zA-Z0-9_.+-]+)@(?P<authority>[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)')  # http://emailregex.com/
+
+    @classmethod
+    def hint(self, obj):
+        if 'mailto:' in obj:
+            return 1.0
+        if '@' in obj:
+            return 0.35
+        return 0.0
+
+    def execute(self, obj):
+        if not isinstance(obj, str):
+            raise TypeError('\'{}\' is not of type str.'.format(obj))
+        emails = self.EMAIL_RE.findall(obj)
+        if len(emails) < 1:
+            raise ValueError('\'{}\'is not a valid email address.'.format(obj))
+        if len(emails) > 1:
+            raise ValueError('Found many email addresses in \'{}\'.'.format(obj))
+        return {
+            'scheme': 'mailto',
+            'authority': emails[0][2],
+            'IRI': 'mailto:{1}@{2}'.format(*emails[0])
+        }
+
+
+class IRILink(AbstractLink):
+    __links = (
```

Comment:
I'm surprised you didn't use `AbstractIRILink.__subclasses__()`. This is fine too, I guess.